splay/src/main/scala/org/scalajs/benchmark/splay/Splay.scala

```diff
@@ -0,0 +1,439 @@
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Google Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// This benchmark is based on a JavaScript log processing module used
+// by the V8 profiler to generate execution time profiles for runs of
+// JavaScript applications, and it effectively measures how fast the
+// JavaScript engine is at allocating nodes and reclaiming the memory
+// used for old nodes. Because of the way splay trees work, the engine
+// also has to deal with a lot of changes to the large tree object
+// graph.
+
+// This is a Scala implementation of the Splay benchmark from:
+// https://github.com/chromium/octane/blob/master/splay.js
+
+// Ported by Nicolas Stucki to Scala.js.
+
+package org.scalajs.benchmark.splay
+
+import scala.scalajs.js
+import scala.util.Random
+
+object Splay extends org.scalajs.benchmark.Benchmark {
+
+  override def prefix: String = "Splay"
+
+  def run(): Unit = splayRun()
+  override def setUp(): Unit = splaySetup()
+  override def tearDown(): Unit = splayTearDown()
+
+  //// Configuration.
+  private final val kSplayTreeSize = 8000
+  private final val kSplayTreeModifications = 80
+  private final val kSplayTreePayloadDepth = 5
+
+  private var splayTree: SplayTree[js.Dynamic] = null
+  private var splaySampleTimeStart = 0L
+
+  private def generatePayloadTree(depth: Int, tag: String): js.Dynamic = {
+    if (depth == 0) {
+      js.Dynamic.literal(
+        array = js.Array(0, 1, 2, 3, 4, 5, 6, 7, 8, 9),
+        string = s"String for key $tag in leaf node"
+      )
+    } else {
+      js.Dynamic.literal(
+        left =  generatePayloadTree(depth - 1, tag),
+        right = generatePayloadTree(depth - 1, tag)
+      )
+    }
+  }
+
+  Random.setSeed(1L)
+  private def generateKey(): Int = {
+    // The benchmark framework guarantees that Math.random is
+    // deterministic; see base.js.
+    Random.nextInt()
```

Comment:
Using `Random.nextInt()` might be a biased source of difference in time that you observe. Make sure to use precisely the same PRNG algorithm on the two benchmarks, otherwise you're skewing the bench on something irrelevant.